File: src/com/xw/vvtts/utils/LanguageSegmenter.java
package com.xw.vvtts.utils;

import java.util.ArrayList;
import java.util.List;

/**
 * 中英混合文本分段路由（移植自广荣 BilingualRoutingPolicy）。
 *
 * 解决"CJK 语音不读英文"的问题：苹果中文/日/韩库只认本国字符，
 * 遇到纯 ASCII 英文/字母段直接丢弃。正确做法不是让中文库硬读英文，
 * 而是按语言切段——中文段走中文引擎，英文段走英文引擎，最后 PCM 拼接。
 *
 * 广荣逻辑（已逆向 routeSegment）：
 *   segmentByLanguage=true 且当前段是英文 → ENGLISH voice
 *   否则 → CHINESE voice
 */
public final class LanguageSegmenter {

    public static class Segment {
        public final String text;
        public final boolean english;  // true=英文段, false=中文段
        public Segment(String t, boolean e) { text = t; english = e; }
    }

    private LanguageSegmenter() {}

    /** 按语言切段。返回的段：中文/符号/数字走中文，纯英文字母串走英文。 */
    public static List<Segment> segment(String text) {
        List<Segment> out = new ArrayList<>();
        if (text == null || text.isEmpty()) return out;

        StringBuilder cur = new StringBuilder();
        int curType = -1;  // 0=中文/其他, 1=英文字母
        for (int i = 0; i < text.length(); i++) {
            char c = text.charAt(i);
            int type = isEnglishChar(c) ? 1 : 0;
            if (curType == -1) {
                curType = type;
            } else if (type != curType) {
                out.add(new Segment(cur.toString(), curType == 1));
                cur = new StringBuilder();
                curType = type;
            }
            cur.append(c);
        }
        if (cur.length() > 0) {
            out.add(new Segment(cur.toString(), curType == 1));
        }
        return out;
    }

    /** 是否为英文字母（a-z A-Z），数字和符号不算英文段（数字交给中文数字读法） */
    private static boolean isEnglishChar(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
    }

    /** 判断某段是否含英文（供上层决定是否整段走英文引擎） */
    public static boolean containsEnglish(String text) {
        if (text == null) return false;
        for (int i = 0; i < text.length(); i++) {
            if (isEnglishChar(text.charAt(i))) return true;
        }
        return false;
    }

    /** 判断某段是否含 CJK（中文/日/韩） */
    public static boolean containsCjk(String text) {
        if (text == null) return false;
        for (int i = 0; i < text.length(); i++) {
            char c = text.charAt(i);
            if ((c >= 0x4E00 && c <= 0x9FFF)      // 基本汉字
                    || (c >= 0x3400 && c <= 0x4DBF) // 扩展A
                    || (c >= 0x3040 && c <= 0x30FF) // 日文假名
                    || (c >= 0xAC00 && c <= 0xD7AF)) { // 韩文
                return true;
            }
        }
        return false;
    }
}

File: src/com/xw/vvtts/utils/LanguageSegmenter.kt
package com.xw.vvtts.utils

/**
 * Routing for mixed Chinese-English text(ported from Guangrong's BilingualRoutingPolicy).
 *
 * Fixes "CJK voices not reading English":Apple's zh/ja/ko banks only recognize native scripts,
 * dropping pure ASCII English/letter runs outright. The right move isn't forcing the zh bank to read English,
 * but splitting by language——zh runs go to the zh engine,,en runs to the en engine,,and PCM is stitched at the end.
 *
 * Guangrong's logic(reverse-engineered from routeSegment):
 *   segmentByLanguage=true and the current run is English -> ENGLISH voice
 *   otherwise -> CHINESE voice
 */
class LanguageSegmenter {
    class Segment(
        val text: String,
       /** true=English run;false=Chinese run */
        val english: Boolean,
    )

    companion object {
       /** Split by language:returned runs——Chinese/symbol/digit go Chinese;pure-English-letter runs go English. */
        fun segment(text: String?): List<Segment> {
            val out = ArrayList<Segment>()
            if (text == null || text.isEmpty()) return out

            var cur = StringBuilder()
            var curType = -1  // 0=Chinese/other, 1=English letter
            for (i in text.indices) {
                val c = text[i]
                val type = if (isEnglishChar(c)) 1 else 0
                if (curType == -1) {
                    curType = type
                } else if (type != curType) {
                    out.add(Segment(cur.toString(), curType == 1))
                    cur = StringBuilder()
                    curType = type
                }
                cur.append(c)
            }
            if (cur.length > 0) {
                out.add(Segment(cur.toString(), curType == 1))
            }
            return out
        }

       /** Whether a run is English letters(a-z A-Z);digitsand symbols don't count(digits go to the Chinese digit reader) */
        private fun isEnglishChar(c: Char): Boolean {
            return (c in 'a'..'z') || (c in 'A'..'Z')
        }

       /** Whether a run contains English(lets callers decide whetherthe whole run goes to the en engine) */
        fun containsEnglish(text: String?): Boolean {
            if (text == null) return false
            for (i in text.indices) {
                if (isEnglishChar(text[i])) return true
            }
            return false
        }

       /** Whether a run contains CJK(Chinese/Japanese/Korean) */
        fun containsCjk(text: String?): Boolean {
            if (text == null) return false
            for (i in text.indices) {
                val c = text[i]
                if ((c.code in 0x4E00..0x9FFF)      // base Han
                        || (c.code in 0x3400..0x4DBF)  // extension A
                        || (c.code in 0x3040..0x30FF)  // Japanese kana
                        || (c.code in 0xAC00..0xD7AF)) { // Hangul
                    return true
                }
            }
            return false
        }
    }
}